### DISCUSSION TITLE
IPXE DHCPV6 Solicit not getting response

### DISCUSSION THREAD

**Discussion by sbucrew** [ipxe/ipxe#739] [General]
Hello IPXE team,

I am fairly new to IPXE and would appreciate your guidance on how to troubleshoot the following issue further. 

Background:
DGX1 on a 40G network with DHCP4 and DCHP6 running.  I am able to acquire IPV6 address from DHCPv6 at UEFI boot which downloads the IPXE image from TFTP successfully.

The IPXE image is configured to run: `ifconf -c ipv6 net4 ` to retrieve an address and then continues with imgfetch and chain loading etc. However it gets stuck at getting a valid IPv6 address. I dropped in to shell and ran it multiple times but I get the same result. (Note: I apologize in advance for changing the mac address 0c:42:ZZ:ZZ:ZZ:50)

`
iPXE> ifconf -c ipv6 net4
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50).................. Connection timed out (http://ipxe.org/4c3b6092)
`
 

```
iPXE> route
...
net4: 100.104.X.248/255.255.255.0 gw 100.X.Y.1 (inaccessible)
net4: 2620:10d:c083:4:e42:ZZff:feZZ:ZZ50/64 gw fe80::200:5eff:fe00:201 (inaccessible)
net4: fe80::e42:ZZff:feZZ:ZZ50/64 (inaccessible)
 
```
 
So above is getting a SLAAC Address w/ local link gw but this is not useful for connectivity out.

I was able to run tcpdump on the switch and compared the DHCPv6 Solicit format from both UEFI boot (successful) and IPXE DHCPv6 request. I noticed the differences in Client Identifier and Vendor Class. Attached 

Uploaded the screenshot of manual comparison for your reference. Is there any other output that I can provide be useful here?

Thanks in advance

![ipxe_dhcpv6_issue_compare](https://user-images.githubusercontent.com/113118749/189152058-17322efe-7654-4d15-9eec-6df875a39178.png)

**Comment by mcb30** [comment]
Firstly, the SLAAC address should be perfectly usable for outbound connections. It's normal in IPv6 for the gateway address to be a link-local address. (The gateway address just needs to be link-local reachable by the client: it doesn't need to be in the same numbered subnet as the source address).

Secondly, for further debugging please try building with DEBUG=dhcpv6 (as suggested on the error page https://ipxe.org/err/4c3b60 shown in your error message)

**Reply by sbucrew** [reply]
Thanks for the debug flag recommendation.. the output highlights that the SOLICIT message going out.. Just curious is there a way to change the Client Identifier to use 'link-layer address' as DUID Type? 

```
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID f002a8
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID bcc315
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID 9e5f10
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID e67870
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID d27114
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
Autoconfiguring net4...
Configuring [ipv6] (net4 0c:42:ZZ:ZZ:ZZ:50)...DHCPv6 net4 has XID 9054a9
DHCPv6 net4 entering SOLICIT state
............... Connection timed out (http://ipxe.org/4c3b6092)
```

**Reply by stappersg** [reply]
That
```text
Connection timed out (http://ipxe.org/4c3b6092)
```
is again http://ipxe.org/4c3b6092 as `mcb30` pointed out.  On that webpage are hints on what could be seen at DHCPv6 server side.  This _discussion_ hasn't yet information about _server side_.
And now is _this update_ a **hint** to add such information.


Regarding the _"is there a way to ..."_ question:  The consensus is that it is unwise to tell it can't be done to those who are doing it.  ( In other words:  I did see that question and could not ignore it completely. )

**Reply by mcb30** [reply]
@sbucrew Sorry for the delay.  It looks very strongly as though your DHCPv6 server is simply not responding to the SOLICIT, and that iPXE is then correctly reporting the server's failure to respond.

There is no way to change the behaviour to use DUID-LL instead of DUID-UUID.  The intention of the [commit that switched to DUID-UUID](https://github.com/ipxe/ipxe/commit/8aab959be) was to match the DUID most likely to be selected by the subsequently booted operating system.  If Linux and Windows now tend to use DUID-LL then we should arguably change to match that.  (Note that we cannot use DUID-LLT, since we have no way to know the timestamp stored or selected by the subsequently booted OS.)  What form of DUID do you see your clients using from within the booted operating system?

**Reply by sbucrew** [reply]
@mcb30 - I appreciate the follow up.  I just booted up in to the OS and manually issued dhclient -6 <interface> and captured it through tcpdump.  (I am running on CentOs7.9, 3.10.0-1127.el7). It looks to be using LinkLayer as well. 

(btw I have a case w/ the switch vendor and expect to hear back this week on why the dhcp relay does not forward the solicit to the dhcp servers. I'll keep this forum posted with details.)

![Screen Shot 2022-09-13 at 11 50 54 AM](https://user-images.githubusercontent.com/113118749/189949239-ef67c4f1-cc1c-450b-86dc-09209f807aec.png)

**Reply by sbucrew** [reply]
@mcb30 -- I got the following response from the vendor.  They've done their analysis and found an extra 14bytes in DHCPV6 header trailer which the router was failing to parse and discarding it. Apparently there was a similar case for the DHCPv4 with IPXE awhile back. https://forum.ipxe.org/printthread.php?tid=11992

Pasting their response below for your reference and they've advised to follow up with IPXE.org .. I'd appreciate your help on this. 

### 


Non working DHCPv6 solicit:

0000  33 33 00 01 00 02 0c 42 a1 1b 2f 50 86 dd 60 00

0010  00 00 00 85 11 ff fe 80 00 00 00 00 00 00 0e 42

0020  a1 ff fe 1b 2f 50 ff 02 00 00 00 00 00 00 00 00

0030  00 00 00 01 00 02 02 22 02 23 00 85 75 27 01 5b  <<<<< here UDP length field is "00 85" == 133

0040  65 fe 00 01 00 12 00 04 31 d1 af b7 2a 36 e9 11

0050  a3 b2 d8 c4 97 d1 98 1b 00 03 00 0c c9 22 90 3a

0060  00 00 00 00 00 00 00 00 00 06 00 08 00 17 00 18

0070  00 3b 00 3c 00 10 00 26 00 00 01 57 00 20 50 58

0080  45 43 6c 69 65 6e 74 3a 41 72 63 68 3a 30 30 30

0090  30 37 3a 55 4e 44 49 3a 30 30 33 30 31 30 00 3d

00a0  00 02 00 07 00 3e 00 03 01 03 0a 00 0f 00 06 00

00b0  04 69 50 58 45 00 08 00 02 00 00 6f 00 75 00 6e

00c0  00 74 00 00 00 14 68 00 74

 

length of dhcp header (a.k.a UDP data ): 139 bytes

length of udp header = 8 byte

UDP length value seen inside the header is 133 which is clearly smaller than the combined values of udp header and udp data (139+8 = 147 bytes)

This is the reason for the error message in dhcp traceoptions: " dhcpv6_option_parse_buffer: Total relay msg len is smaller than embedded options len"

===========

As per engineering team, It is clear that in non-working case there are extra 14bytes in DHCPV6 header trailer. 

Today in our LAB we removed these extra bytes 14bytes (6f 00 75 00 6e 00 74 00 00 00 14 68 00 74) from DHCPV6 header trailer and replayed the packet. This packet successfully got forwarded by DHCP Relay agent.

To find out if this is miscalculated length or extra bytes we increased the IPv6 header payload and UDP header length by 14 to accommodate these extra 14 bytes. However wireshark was not able to get meaning of these extra 14 bytes, so this looks not part of original DHCP header. It looks iPXE NIC is adding these extra 14 bytes later. 

There is similar article in iPXE.org for DHCPV4 where same 14 extra bytes is being add in the end and it was causing issue with DHCP relay.  https://forum.ipxe.org/printthread.php?tid=11992

Engineering recommends checking with vendor or ipxe.org to find out what is these extra bytes and how to resolve this.

Juniper device is working as expected in this case.

**Reply by sbucrew** [reply]
Update - I did some analysis and looks like the vendor is wrong. They are accounting the trailer/FCS as part of the payload. I punted it back to them. 

![analysis copy](https://user-images.githubusercontent.com/113118749/191540134-d5555dac-3504-4024-b697-5af70756a09f.jpg)

**Reply by stappersg** [reply]
> looks like the vendor is wrong

The vendor of what?

**Reply by sbucrew** [reply]
So the  network switch vendor confirmed their bug on how they parse the payload. I am looking at a workaround as the fix for that might not be coming in soon..  Potentially excluding (trailer 10bytes + fcs 4bytes) from the packet structure. I am searching for the code where these get added but any guidance would be appreciated

**Reply by stappersg** [reply]
Acknowledge on _"network switch vendor"_.   About
> guidance would be appreciated

Make it possible that there is a win-win-situation.  So benefits for those who got a problem solved **and** benefits for those who did contribute to solve the problem. Think ["How to keep iPXE in good shape"](https://github.com/ipxe/ipxe/discussions/725).

**Reply by sbucrew** [reply]
@stappersg - please stop with your spam and useless comments that are polluting this thread.

**Reply by mcb30** [reply]
@subcrew The fact that it's 14 bytes suggests that something in the stack is getting confused between "packet length (including 14-byte Ethernet frame header)" and "payload length (excluding 14-byte Ethernet frame header)".

There was a similar issue addressed in commit https://github.com/ipxe/ipxe/commit/83b8c0e21.  I don't think you've mentioned which iPXE version you are using, so please check that you are using the most recent version.

I've verified that iPXE's NII and SNP drivers transmit packets of the correct total length when using either the VirtioNetDxe driver from EDK2 or the Intel UEFI driver.  This tells me that the current behaviour in iPXE is almost certainly correct, and that if the problem persists that it is most likely to be a bug in the Mellanox driver provided by your firmware.

**Reply by sbucrew** [reply]
@mcb30  - thank you for your response and appreciate the details. Yes it was indeed that was the issue! Apparently we had a very old copy in our internal repo. I traced and mod'd the code where that add'l 14bytes was introduced and got it to work. I was going post an update back here in Oct but a spammer has spoiled the thread.

**Comment by stappersg** [comment]
For future discussion:

![afbeelding](https://user-images.githubusercontent.com/35434094/191943220-029ffab2-42a7-4bf2-9af4-ca1125e6a0bd.png)
